File: leetcode/leetcode_py/find-peak-element.py
'''
https://leetcode.com/problems/find-peak-element

Solution1. linear
time: o(n)
space: o(1)

Solution2. binary search
time: o(logn)
space: o(1)
'''
class Solution:
    def findPeakElement_linear(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1: return 0
        
        maxValue = nums[0]
        maxIdx = 0
        for i in range(1, n):
            if maxValue < nums[i]:
                maxValue = nums[i]
                maxIdx = i
        return maxIdx

    def findPeakElement_bs(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1: return 0
        l = 0
        r = n - 1
        while l < r:
            m = l + (r - l) // 2
            if nums[m] > nums[m+1]:
                r = m
            else:
                l = m+1
            
        return l
        
    def findPeakElement(self, nums: List[int]) -> int:
        return self.findPeakElement_bs(nums)
            
        

File: leetcode/leetcode_cpp/find-peak-element.cpp
/*
https://leetcode.com/problems/find-peak-element

Solution1. linear
time: o(n)
space: o(1)

Solution2. binary search
time: o(logn)
space: o(1)

*/
class Solution {
public:
    int findPeakElement_linear(vector<int>& nums) {
        for (int i=0; i<nums.size()-1; ++i)
            if (nums[i] > nums[i+1]) return i;
        return nums.size()-1;
    }
    int findPeakElement_bs(vector<int>& nums) {
        int n = nums.size();
        int low = 0;
        int high = nums.size()-1;
        while(low < high){
            int mid = low + (high - low) / 2;
            if(nums[mid] > nums[mid+1]){
                high = mid;
            }
            else{
                low = mid + 1;
            }
        }
        return low;
    }
    int findPeakElement(vector<int>& nums) {
        return findPeakElement_bs(nums);
    }
};